File: experiments/data_process/gather_data.jl
#!/usr/bin/env julia

using Printf
using RobotOS
using Dates
using DataFrames
using CSV

# include("unscented_transform.jl")

@rosimport mrs_msgs.msg:RangeWithCovarianceArrayStamped, ControlManagerDiagnostics, PoseWithCovarianceArrayStamped
@rosimport geometry_msgs.msg:PoseWithCovarianceStamped
@rosimport nav_msgs.msg:Odometry
rostypegen()

global data = DataFrame(
    timestamp=Float64[],
    range_gt=Float64[],
    range_ot=Float64[],
    range_var_ot=Float64[],
    mh=Float64[],
    x_gt=Float64[],
    y_gt=Float64[],
    z_gt=Float64[],
    x_ot=Float64[],
    y_ot=Float64[],
    z_ot=Float64[],
    x_obs=Float64[],
    y_obs=Float64[],
    z_obs=Float64[],
)

global distance_data = DataFrame(
    timestamp=Float64[], 
    range_uwb=Float64[], 
    variance=Float64[],
    rtk_distance=Float64[],
    angle=Float64[]
)

global uav42_rtk = [0, 0, 0]
global uav37_rtk = [0, 0, 0]

global START_TIME::Float64 = NaN

function object_tracker_cb(msg::mrs_msgs.msg.PoseWithCovarianceArrayStamped)
    try
        time_stamp::Float64 = (convert(Float64, msg.header.stamp.secs) + msg.header.stamp.nsecs * 1e-9)
        global START_TIME

        @printf "Got data from object tracker\n\r"

        if (isnan(START_TIME) || time_stamp < START_TIME)
            START_TIME = time_stamp
        end
    
        time_stamp = time_stamp - START_TIME

        uav1 = position_to_vector(uav1_gt.pose.pose.position)
        position_gt_g = position_to_vector(uav2_gt.pose.pose.position)

        position_gt_r = position_gt_g - uav1
        ground_truth_dist = (position_gt_r' * position_gt_r)^(1 / 2)

        uav1_q = uav1_gt.pose.pose.orientation
        quaternion = quat(uav1_q.w, uav1_q.x, uav1_q.y, uav1_q.z)

        for measurement in msg.poses
            position_r_raw = position_to_vector(measurement.pose.position)
            position_r = rotate_vector(conj(quaternion), position_r_raw)
            position_g = position_r + uav1
            range = (position_r' * position_r)^(1 / 2)

            covariance_raw = reshape(measurement.covariance, (6, 6))
            covariance_raw = covariance_raw'
            covariance_raw = covariance_raw[1:3, 1:3]
            covariance = rotmatrix_from_quat(conj(quaternion)) * covariance_raw
            mahanalobis = sqrt((position_gt_r - position_r)' * inv(covariance) * (position_gt_r - position_r))

            σ, Wm, Wc = UT.compute_sigma_pts(position_r_raw, covariance_raw)

            @. σ ^= 2 
            σ = sum(σ, dims=2)
            @. σ ^= 1/2

            μ = Wm'*σ
            Σ = zeros(Float64, (1, 1))

            for (index, value) in enumerate(Wc)
                @. Σ += value*(σ[index, :] - μ)*(σ[index, :] - μ)'
            end

            push!(data, [time_stamp 
            ground_truth_dist 
            μ[1, 1] 
            Σ[1, 1] 
            mahanalobis 
            position_gt_g...
            position_g...])
            # filter!(row -> row.timestamp > time_stamp - 20, data[measurement.id])
        end

        CSV.write("object_tracker.csv", data)
    catch y
        println(y)
        rethrow(y)
    end
end


function uwb_cb(msg::mrs_msgs.msg.RangeWithCovarianceArrayStamped)
    time_stamp::Float64 = (convert(Float64, msg.header.stamp.secs) + msg.header.stamp.nsecs * 1e-9)
    global START_TIME
    global distance_data

    global uav37_rtk
    global uav42_rtk

    angle = atan(uav37_rtk[2] - uav42_rtk[2], uav37_rtk[1] - uav42_rtk[1])

    rtk_distance::Float64 = sqrt((uav42_rtk - uav37_rtk)' * (uav42_rtk - uav37_rtk))

    if rtk_distance < 0.1 || rtk_distance > 200
        return
    end

    if (isnan(START_TIME) || time_stamp < START_TIME)
        START_TIME = time_stamp
    end

    time_stamp = time_stamp - START_TIME

    # rtk_distance::Float64 = 0

    for measurement in msg.ranges
        if (measurement.range.range < 0)
            return
        end

        push!(distance_data, [time_stamp measurement.range.range measurement.variance rtk_distance angle])
    end
end

function ground_truth_cb(msg::nav_msgs.msg.Odometry, uav::Int)
    global uav37_rtk
    global uav42_rtk
    global distance_data

    if uav == 1
        uav37_rtk = [msg.pose.pose.position.x, msg.pose.pose.position.y, msg.pose.pose.position.z]
    end
    if uav == 2
        uav42_rtk = [msg.pose.pose.position.x, msg.pose.pose.position.y, msg.pose.pose.position.z]
    end
end

function uav37_odom_cb(msg::nav_msgs.msg.Odometry)
    ground_truth_cb(msg, 1)
end

function uav42_odom_cb(msg::nav_msgs.msg.Odometry)
    ground_truth_cb(msg, 2)
end

function save_csv()
    global distance_data
    @printf "Saving data\n\r"
    CSV.write("data.csv", distance_data)
end

function main()
    global distance_data

    init_node("gather_data", anonymous=true)

    CSV.write("object_tracker.csv", data)
    CSV.write("data.csv", distance_data)

    @printf "Gather data\n\r"

    t = Timer((t) -> save_csv(), 1; interval=1) 

    uav37_odom_sub = Subscriber{nav_msgs.msg.Odometry}("/uav37/estimation_manager/rtk/odom", uav37_odom_cb)
    uav42_odom_sub = Subscriber{nav_msgs.msg.Odometry}("/uav42/estimation_manager/rtk/odom", uav42_odom_cb)

    object_tracker_sub = Subscriber{mrs_msgs.msg.PoseWithCovarianceArrayStamped}("/uav42/object_tracker/filtered_poses", object_tracker_cb)
    uwb_sub = Subscriber{mrs_msgs.msg.RangeWithCovarianceArrayStamped}("/uav42/uwb_range/range", uwb_cb)

    while(true)
      try
        spin()
      catch y
        # nothing
      end
    end
end

main()

File: experiments/data_process/gather_data.py
import h5py
import numpy as np
import rospy
import argparse

import message_filters
from sensor_msgs.msg import Image, CameraInfo
from geometry_msgs.msg import TwistStamped
from mrs_msgs.msg import PoseWithCovarianceArrayStamped, RangeWithCovarianceArrayStamped, UavState
from nav_msgs.msg import Odometry

file = h5py.File('data.h5', 'w')

file.create_group('object_tracker')
file.create_group('object_tracker/filtered_poses')
file.create_group('uwb')

uwb = None

rtk_leader = None
rtk_follower = None

rtk_leader_velocity = None

def object_tracker_callback(fcu: PoseWithCovarianceArrayStamped, rtk: PoseWithCovarianceArrayStamped, velocity: TwistStamped):
    global file
    global rtk_leader
    global rtk_follower

    if(len(fcu.poses) < 1 or len(rtk.poses) < 1):
        return

    if(rtk_leader is None or rtk_follower is None):
        return

    if(not '/object_tracker/timestamp' in file):
        file['object_tracker'].create_dataset('timestamp', data=[fcu.header.stamp.to_sec()], maxshape=(None,))
    else:
        file['object_tracker/timestamp'].resize((file['object_tracker/timestamp'].shape[0] + 1), axis=0)
        file['object_tracker/timestamp'][-1] = fcu.header.stamp.to_sec()
    
    if(not '/object_tracker/leader' in file):
        file['object_tracker'].create_dataset('leader', data=rtk_leader, maxshape=(None,7))
    else:
        file['object_tracker/leader'].resize((file['object_tracker/leader'].shape[0] + 1), axis=0)
        file['object_tracker/leader'][-1] = rtk_leader

    if(not '/object_tracker/follower' in file):
        file['object_tracker'].create_dataset('follower', data=rtk_follower, maxshape=(None,7))
    else:
        file['object_tracker/follower'].resize((file['object_tracker/follower'].shape[0] + 1), axis=0)
        file['object_tracker/follower'][-1] = rtk_follower

    poses = {"fcu": fcu.poses[0], "rtk": rtk.poses[0]}

    if(not '/object_tracker/velocity_truth' in file):
        file['object_tracker'].create_dataset('velocity_truth', data=rtk_leader_velocity, maxshape=(None,3))
    else:
        file['object_tracker/velocity_truth'].resize((file['object_tracker/velocity_truth'].shape[0] + 1), axis=0)
        file['object_tracker/velocity_truth'][-1] = rtk_leader_velocity

    if(not '/object_tracker/velocity' in file):
        file['object_tracker'].create_dataset('velocity', data=np.array([velocity.twist.linear.x, velocity.twist.linear.y, velocity.twist.linear.z]).reshape((1,3)), maxshape=(None,3))
    else:
        file['object_tracker/velocity'].resize((file['object_tracker/velocity'].shape[0] + 1), axis=0)
        file['object_tracker/velocity'][-1] = np.array([velocity.twist.linear.x, velocity.twist.linear.y, velocity.twist.linear.z]).reshape((1,3))
    
    for key, pose in poses.items():
        if( not '/object_tracker/filtered_poses/' + key in file):
            file['object_tracker/filtered_poses'].create_group(key)
            file['object_tracker/filtered_poses/' + key].create_dataset('position', data=np.array([pose.pose.position.x, pose.pose.position.y, pose.pose.position.z]).reshape((1, 3)),  maxshape=(None,3))
            file['object_tracker/filtered_poses/' + key].create_dataset('orientation', data=np.array([pose.pose.orientation.x, pose.pose.orientation.y, pose.pose.orientation.z, pose.pose.orientation.w]).reshape((1,4)), maxshape=(None,4))
        else:
            file['object_tracker/filtered_poses/' + key + '/position'].resize((file['object_tracker/filtered_poses/' + key + '/position'].shape[0] + 1), axis=0)
            file['object_tracker/filtered_poses/' + key + '/position'][-1] = np.array([pose.pose.position.x, pose.pose.position.y, pose.pose.position.z]).reshape((1, 3))
            file['object_tracker/filtered_poses/' + key + '/orientation'].resize((file['object_tracker/filtered_poses/' + key + '/orientation'].shape[0] + 1), axis=0)
            file['object_tracker/filtered_poses/' + key + '/orientation'][-1] = np.array([pose.pose.orientation.x, pose.pose.orientation.y, pose.pose.orientation.z, pose.pose.orientation.w]).reshape((1,4))

        if not '/object_tracker/filtered_poses/' + key + '/covariance' in file:
            file['object_tracker/filtered_poses/' + key].create_dataset('covariance', data=np.array(pose.covariance).reshape((1,36)), maxshape=(None, 36))
        else:
            file['object_tracker/filtered_poses/' + key + '/covariance'].resize((file['object_tracker/filtered_poses/' + key + '/covariance'].shape[0] + 1), axis=0)
            file['object_tracker/filtered_poses/' + key + '/covariance'][-1] = np.array(pose.covariance).reshape((1,36))
    return

def uwb_callback(range: RangeWithCovarianceArrayStamped):
    global file
    global rtk_leader
    global rtk_follower

    if len(range.ranges) < 1:
        return
    
    if(rtk_leader is None or rtk_follower is None):
        return

    if(not '/uwb/timestamp' in file):
        file['uwb'].create_dataset('timestamp', data=[range.header.stamp.to_sec()], maxshape=(None,))
    else:
        file['uwb/timestamp'].resize((file['uwb/timestamp'].shape[0] + 1), axis=0)
        file['uwb/timestamp'][-1] = range.header.stamp.to_sec()

    if(not '/uwb/leader' in file):
        file['uwb'].create_dataset('leader', data=rtk_leader, maxshape=(None,7))
    else:
        file['uwb/leader'].resize((file['uwb/leader'].shape[0] + 1), axis=0)
        file['uwb/leader'][-1] = rtk_leader

    if(not '/uwb/follower' in file):
        file['uwb'].create_dataset('follower', data=rtk_follower, maxshape=(None,7))
    else:
        file['uwb/follower'].resize((file['uwb/follower'].shape[0] + 1), axis=0)
        file['uwb/follower'][-1] = rtk_follower

    if(not '/uwb/range' in file):
        file['uwb'].create_dataset('range', data=np.array(range.ranges[0].range.range).reshape((1,1)), maxshape=(None,1))
        file['uwb'].create_dataset('covariance', data=np.array(range.ranges[0].variance).reshape((1,1)), maxshape=(None,1))
        file['uwb'].create_dataset('power', data=np.array((range.ranges[0].power_a + range.ranges[0].power_b)/2).reshape((1,1)), maxshape=(None,1))
    else:
        file['uwb/range'].resize((file['uwb/range'].shape[0] + 1), axis=0)
        file['uwb/range'][-1] = np.array(range.ranges[0].range.range).reshape((1,1))
        file['uwb/covariance'].resize((file['uwb/covariance'].shape[0] + 1), axis=0)
        file['uwb/covariance'][-1] = np.array(range.ranges[0].variance).reshape((1,1))
        file['uwb/power'].resize((file['uwb/power'].shape[0] + 1), axis=0)
        file['uwb/power'][-1] = np.array((range.ranges[0].power_a + range.ranges[0].power_b)/2).reshape((1,1))

    return

def rtk_leader_callback(odom: UavState):
    global rtk_leader
    global rtk_leader_velocity
    rtk_leader_velocity = np.array([odom.velocity.linear.x, odom.velocity.linear.y, odom.velocity.linear.z]).reshape((1,3))
    rtk_leader = np.array([odom.pose.position.x, 
                           odom.pose.position.y, 
                           odom.pose.position.z, 
                           odom.pose.orientation.x, 
                           odom.pose.orientation.y, 
                           odom.pose.orientation.z, 
                           odom.pose.orientation.w]).reshape((1,7))

def rtk_follower_callback(odom: UavState):
    global rtk_follower
    rtk_follower = np.array([odom.pose.position.x, 
                           odom.pose.position.y, 
                           odom.pose.position.z, 
                           odom.pose.orientation.x, 
                           odom.pose.orientation.y, 
                           odom.pose.orientation.z, 
                           odom.pose.orientation.w]).reshape((1,7))

def main():
    global file
    global rtk_leader
    global rtk_follower
    global uwb

    parser = argparse.ArgumentParser()
    parser.add_argument('file', help='file name to save data to')
    args = parser.parse_args()

    file = h5py.File(args.file, 'w')

    file.create_group('object_tracker')
    file.create_group('object_tracker/filtered_poses')
    file.create_group('uwb')

    rospy.init_node('gather_data')

    object_tracker_fcu_sub = message_filters.Subscriber('/uav42/object_tracker_offline/filtered_poses_fcu', PoseWithCovarianceArrayStamped)
    object_tracker_rtk_sub = message_filters.Subscriber('/uav42/object_tracker_offline/filtered_poses_rtk', PoseWithCovarianceArrayStamped)
    object_tracker_velocity_sub = message_filters.Subscriber('/uav42/object_tracker_offline/filtered_velocity', TwistStamped)

    rtk_leader_sub = rospy.Subscriber('/uav37/estimation_manager/rtk/uav_state', UavState, rtk_leader_callback)
    rtk_follower_sub = rospy.Subscriber('/uav42/estimation_manager/rtk/uav_state', UavState, rtk_follower_callback)

    uwb_sub = rospy.Subscriber('/uav42/uwb_range/range', RangeWithCovarianceArrayStamped, uwb_callback)

    ts = message_filters.TimeSynchronizer([object_tracker_fcu_sub, object_tracker_rtk_sub, object_tracker_velocity_sub], 10)
    ts.registerCallback(object_tracker_callback)

    rospy.spin()

if __name__ == '__main__':
    main()
